i18nの提出物

yocaji opened the pull request:
i18nの提出物になります。レビューよろしくお願いいたします🙏

Commit:
i18n_generatorsを入れた

- rails g i18n ja
- rails g i18n en

Commit:
i18nに対応した

- 辞書ファイル作成
- view書き換え
- クエリパラメーターによるlocale切り替え機能を追加

Commit:
Gemfileの書式を修正した

Commit:
使っていないテンプレート変数を削除した

cafedomancer commented on app/views/books/index.html.erb:
```diff
@@ -1,14 +1,17 @@
+<%= link_to '日本語', './books?locale=ja' %>
+<%= link_to 'English', './books?locale=en' %>
```
リンク先のパスには named route helper が使えると思うのでやってみてください！

yocaji replied:
修正しました。`_path`や`_url`という便利なヘルパーメソッドがあるという事、勉強になりました。
（`named route helper`の認識が違ってたらすみません）

cafedomancer replied:
認識あっています！

cafedomancer commented on app/views/books/index.html.erb:
```diff
@@ -1,14 +1,17 @@
-      <th>Memo</th>
-      <th>Author</th>
-      <th>Picture</th>
+      <th><%= t Book.human_attribute_name('title') %></th>
+      <th><%= t Book.human_attribute_name('memo') %></th>
+      <th><%= t Book.human_attribute_name('author') %></th>
+      <th><%= t Book.human_attribute_name('picture') %></th>
```
`human_attribute_name` は明示的に translate しなくても自動的に translate してくれそうです 👀 Rails Guides を確認してみてください。https://guides.rubyonrails.org/i18n.html#translations-for-active-record-models

ちなみに `bin/rails console` で同じコードを実行すると translation missing になってしまいました。これは `I18n.t 'タイトル'` となってしまうからだと思います 😲 

```
irb(main):009:0> I18n.t Book.human_attribute_name('title')
=> "translation missing: ja.タイトル"
```

`show.erb.html` のほうも同様の対応をおねがいします 🙏

yocaji replied:
`rails c`での確認を試して、自動的にtranslateされていることが理解できました🙇‍♀️
修正しました。

cafedomancer commented on config/initializers/locale.rb:
```diff
@@ -0,0 +1,2 @@
+I18n.available_locales = [:en, :ja]
+I18n.default_locale = :ja
```
参考情報: Rails の config として設定することもできます 💁 https://guides.rubyonrails.org/configuring.html#configuring-i18n (i18n ライブラリに delegate していると書いているのでどちらで設定してもよさそうではあります)

yocaji replied:
せっかくなので教えていただいた設定方法に変更しました。

cafedomancer commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,9 @@
+  private
+
+  def default_url_options(options = {})
+    { locale: I18n.locale }.merge options
```
(参考情報なので修正しなくても OK) `reverse_merge` というメソッドを使うとこんなふうに書けます 💁 デフォルト値を設定するときによく使うメソッドです。`options` を主体にできるので自然なコードにできます。 https://api.rubyonrails.org/classes/Hash.html#method-i-reverse_merge

```suggestion
    options.reverse_merge(locale: I18n.locale)
```

yocaji replied:
せっかくなので教えていただいた設定方法に変更しました。
確かにoptionsから始められると読みやすいです👀✨

Commit:
レビューの指摘事項に対応した

- 言語切り替えのリンクにURLヘルパーを利用するようにした
- human_attribute_nameに対する余計なtransrate処理を削除
- デフォルトロケールの設定をRailsのconfigで行うようにした
- パラメータのデフォルト値を設定する処理をmergeからreverse_mergeに変更

cafedomancer reviewed (approved):
🌟
